### ISSUE TITLE
Inventory numbers and table entries

### ISSUE THREAD

**Issue by kadowa** [kadowa/papyri#9]
We need a unique project name to identify papyri in DWork, so I'm looking at what (I suppose) are the papyri inventory numbers. At the moment, I am pretty confused.

The main problem – for me – is that it is not clear, what a record in the table corresponds to. A papyrus? Recto or verso of a papyrus? A part of a papyrus in a certain language? A publication on a papyrus (or part of it)? I suspect that actually all of these occur.

Furthermore, I have trouble understanding how inventory numbers are composed and what they are assigned to (physical objects or catalogue entries?).

The sample catalogue entry in DWork uses an identifier ```p_heid_inv_g_1355```. I guess this is a combination of the fields ```p_heid_inv``` and ```p_heid_inv_nr```. But even taking all fields with the prefix ```p_heid_inv```/```p.heid.inv.``` into account, identifiers generated from these fields would not be unique for every entry in the database:

![p_heid_inv](https://cloud.githubusercontent.com/assets/8066366/12675920/1f38504e-c68e-11e5-959b-62818c0000d3.png)

[source](http://heidi.ub.uni-heidelberg.de/papyri/default/table?keywords=&order=papyri.p_heid_inv_nr)

For example, there are three entries for *Aram.* with ```p_heid_inv_nr``` *1* ([4977](http://heidi.ub.uni-heidelberg.de/papyri/default/table/view/papyri/4977), [4978](http://heidi.ub.uni-heidelberg.de/papyri/default/table/view/papyri/4978) and [8192](http://heidi.ub.uni-heidelberg.de/papyri/default/table/view/papyri/8192) plus one with 'Aram. (arab)', [8194](http://heidi.ub.uni-heidelberg.de/papyri/default/table/view/papyri/8194)). 4977 and 4978 seem to be recto and verso, but what is 8192? It seems similar to 4977, but ```Art 1``` differs (*literarisch* vs. *dokumentarisch*) and also the measures (```breite```/```hoehe```).

@jcowey, to gain a better understanding of the hidden structures in the table it would be very helpful if you could provide me with a detailed description of how you assign inventory numbers and **what** it is that is supposed to form a catalogue entry.

As a consequence, we might have to radically restructure the database: one table which actually represents every physical papyrus linked to from another table for the separate parts (recto/verso/separate letters or languages or publications etc.). Fields would need to be split between the two (or more?) tables; the unifying elements would be the inventory number. This might get ugly, since I have the impression that the current table is a mix of different structures; some papyri have one entry for both recto and verso, some have separate entries (and some have more than that).

**Comment by jcowey** [comment]
Please try building the inventory numbers in the following manner:  
"P.Heid. Inv." + content of "P.Heid. Inv." +  content of "P_Heid_Inv_Nr" +  content of "P.Heid. Inv. Buchstabe" +  content of "RV" + content of "P_Heid_Inv_Erg"  

and if any of the above fields has the content "None" then simply leave that value blank and go to the next one.

This method will replace what was used in the FileMaker database to produce P_Heid_Inv_L (_a field that we got did not read into the new MySQL database_) (= P.Heid. Inv. - lang, d.h. alles was zu dem Inventarnummer gehört). The internal FileMaker script was:  
P.Heid. Inv. & " " & LiesAlsZahl(P_Heid_Inv_Nr) & Wenn(P.Heid. Inv. Buchstabe ≠ "";P.Heid. Inv. Buchstabe;"") & Wenn(RV ≠ ""; " " & RV; "") & " " & P_Heid_Inv_Erg

**Comment by jcowey** [comment]
If we are lucky this might give us "unique identifiers"

**Comment by kadowa** [comment]
Ok, I just checked the resulting inventory numbers (we don't need to store them in a separate field as long as we know how to generate them). Here is a list of the non-unique offenders (the order of columns is ```p_heid_inv	p_heid_inv_nr	p_heid_inv_buchstabe	rv	p_heid_inv_erg```):

```
Arab.	1598	NULL	Recto	NULL
Arab.	2309	NULL	Vorderseite	NULL
Arab.	2386	NULL	Vorderseite	NULL
Arab.	2765	NULL	Vorderseite	NULL
Arab.	468	NULL	Vorderseite u. Rückseite	NULL
Arab.	50	NULL	Recto u. Verso	+ Arab. 51 + Arab. 52 + Arab. 53a + Arab. 53b + Arab. 53c + Arab. 53d + Arab. 53e
Aram.	1	NULL	Vorderseite	NULL
G	1289	NULL	Recto	NULL
G	1368	NULL	Recto	NULL
G	1517	NULL	Vorderseite	NULL
G	4737	NULL	Verso	NULL
G	4745	NULL	Verso	NULL
Hebr.	15	NULL	Vorderseite	NULL
Hebr.	17	NULL	Vorderseite	NULL
Hebr.	18	NULL	Vorderseite	NULL
Hebr.	19	NULL	Vorderseite	NULL
Hebr.	20	NULL	Vorderseite	NULL
Hebr.	22	NULL	Vorderseite	NULL
Hebr.	24	NULL	Vorderseite	NULL
Hebr.	25	NULL	Vorderseite	NULL
Kopt.	1207	NULL	Recto	NULL
Kopt.	12	NULL	Rückseite	Frg. 1-7 (Taf. 1-3)
Kopt.	12	NULL	Vorderseite	Frg. 1-7 (Taf. 1-3)
Kopt.	1349	NULL	Recto	NULL
Kopt.	169	NULL	Recto	NULL
Kopt.	452	NULL	Haarseite u. Fleischseite	NULL
Kopt.	685	NULL	Recto u. Verso	NULL
Kopt.	998	NULL	Recto	NULL
```
The example I detailed above, _Aram_ _1_ is among them. You'll probably want to check these entries manually to find out what is going on there. Luckily, the list seems manageable.

PS: I'm very curious; what is 'Haarseite u. Fleischseite'? :smile:

**Comment by jcowey** [comment]
"Haarseite u. Fleischseite". You have to think in terms of "Pergament". It really is the side where the remaining hair was scraped off, or where the remaining bits of flesh tissue were removed to leave the parchment in perfect condition for writing (keine Krümelchen, keine Borste!!!). :)

**Comment by jcowey** [comment]
Thanks for the list. That should indeed be manageable.

**Comment by kadowa** [comment]
:sheep: :goat: :arrow_right: :scroll:  :grin:

Quite fascinating, thanks!

**Comment by kadowa** [comment]
Sorry, one more off-topic question. I just realized: this is the ***papyri*** database. What's a ***pergament*** doing in there?  :see_no_evil:

**Comment by kadowa** [comment]
Here is the updated list on non-unique identifiers:

```
Arab.	1598	NULL	Recto
Arab.	2309	NULL	Vorderseite
Arab.	2386	NULL	Vorderseite
Arab.	2765	NULL	Vorderseite
Arab.	468	NULL	Vorderseite u. Rückseite
Arab.	50	NULL	Recto u. Verso
Arab.	7	NULL	Recto
Aram.	1	NULL	Vorderseite
G	1289	NULL	Recto
G	1368	NULL	Recto
G	14	NULL	Recto
G	1517	NULL	Vorderseite
G	1744	NULL	Recto
G	1752	NULL	Recto
G	216	NULL	Recto
G	22	NULL	Recto
G	310	NULL	Recto
G	32	NULL	Recto
G	4737	NULL	Verso
G	4745	NULL	Verso
G	81	NULL	Recto
G	87	NULL	Recto u. Verso
G	8	NULL	Recto
Hebr.	15	NULL	Vorderseite
Hebr.	17	NULL	Vorderseite
Hebr.	18	NULL	Vorderseite
Hebr.	19	NULL	Vorderseite
Hebr.	20	NULL	Vorderseite
Hebr.	22	NULL	Vorderseite
Hebr.	24	NULL	Vorderseite
Hebr.	25	NULL	Vorderseite
Kopt.	1207	NULL	Recto
Kopt.	12	NULL	Rückseite
Kopt.	12	NULL	Vorderseite
Kopt.	1349	NULL	Recto
Kopt.	169	NULL	Recto
Kopt.	452	NULL	Haarseite u. Fleischseite
Kopt.	685	NULL	Recto u. Verso
Kopt.	998	NULL	Recto
```

**Comment by jcowey** [comment]
Many thanks. 
Re your question above "Sorry, one more off-topic question. I just realized: this is the papyri database. What's a pergament doing in there?"
Papyri is a rather loosely used term. We include pergament, wood and even bone!!
One of my favourites is:
http://heidi.ub.uni-heidelberg.de/papyri/default/table/view/papyri/1712. :)

**Comment by jcowey** [comment]
- [ ] Arab.	1598	NULL	Recto
- [ ] Arab.	2309	NULL	Vorderseite
- [ ] Arab.	2386	NULL	Vorderseite
- [ ] Arab.	2765	NULL	Vorderseite
- [ ] Arab.	468	NULL	Vorderseite u. Rückseite
- [ ] Arab.	50	NULL	Recto u. Verso
- [ ] Arab.	7	NULL	Recto
- [ ] Aram.	1	NULL	Vorderseite
- [ ] G	1289	NULL	Recto
- [ ] G	1368	NULL	Recto
- [ ] G	14	NULL	Recto
- [ ] G	1517	NULL	Vorderseite
- [ ] G	1744	NULL	Recto
- [ ] G	1752	NULL	Recto
- [ ] G	216	NULL	Recto
- [ ] G	22	NULL	Recto
- [ ] G	310	NULL	Recto
- [ ] G	32	NULL	Recto
- [ ] G	4737	NULL	Verso
- [ ] G	4745	NULL	Verso
- [ ] G	81	NULL	Recto
- [ ] G	87	NULL	Recto u. Verso
- [ ] G	8	NULL	Recto
- [ ] Hebr.	15	NULL	Vorderseite
- [ ] Hebr.	17	NULL	Vorderseite
- [ ] Hebr.	18	NULL	Vorderseite
- [ ] Hebr.	19	NULL	Vorderseite
- [ ] Hebr.	20	NULL	Vorderseite
- [ ] Hebr.	22	NULL	Vorderseite
- [ ] Hebr.	24	NULL	Vorderseite
- [ ] Hebr.	25	NULL	Vorderseite
- [ ] Kopt.	1207	NULL	Recto
- [ ] Kopt.	12	NULL	Rückseite
- [ ] Kopt.	12	NULL	Vorderseite
- [ ] Kopt.	1349	NULL	Recto
- [ ] Kopt.	169	NULL	Recto
- [ ] Kopt.	452	NULL	Haarseite u. Fleischseite
- [ ] Kopt.	685	NULL	Recto u. Verso
- [ ] Kopt.	998	NULL	Recto

**Comment by jcowey** [comment]
Arab. 2309 bei der Bildzuweisung aufpassen.

**Comment by jcowey** [comment]
Arab. 2386 bei der Bildzuweisung aufpassen.

**Comment by jcowey** [comment]
Arab. 2765 bei der Bildzuweisung aufpassen.

**Comment by jcowey** [comment]
G 1517 was catalogued under "G" = Greek. It is, in fact, a Coptic document. This object was catalogued once in the Greek Filemaker Database AND once in the Coptic Filemaker Database which led to us having TWO records in the initial crosswalk to the MySQL database.

**Comment by jcowey** [comment]
G 1744 a literary piece AND a documentary piece on the Recto. The Verso is blank. Is it not the case the the value A in 13415 and B in 13416 provide a unique id? Otherwise this is a difficult case.

**Comment by jcowey** [comment]
G 1752 a receipt AND a different receipt (written by different hands) on the Recto. The Verso is blank. Is it not the case the the value A in 13427 and B in 13428 provide a unique id? Otherwise this is a difficult case.

**Comment by jcowey** [comment]
If you look at the comments on G 1744 and G 1752 above, you will see that I have a question about the entries in the field `p_heid_inv_buchstabe`

**Comment by kadowa** [comment]
@jcowey, sorry, there were a lot of comments on this project lately, so I just skim read them. :wink: If you have an urgent question, you can address me directly in the respective comment with and @-mention.

**Comment by jcowey** [comment]
ok done
@kadowa

**Comment by kadowa** [comment]
You could also use emojis to make questions stand out, (e.g. :question:, for more options http://www.emoji-cheat-sheet.com/); otherwise it can be hard to spot them, if there are many comments on a thread.

**Comment by kadowa** [comment]
@jcowey, you are completely right – I forgot to add the column `p_heid_inv_erg` when I produced the results above. Here is the updated list of non-unique ids:

```
Arab.	1598	NULL	NULL	Recto
Arab.	2309	NULL	NULL	Vorderseite
Arab.	2386	NULL	NULL	Vorderseite
Arab.	2765	NULL	NULL	Vorderseite
Arab.	468	NULL	NULL	Vorderseite u. Rückseite
Arab.	50	NULL	+ Arab. 51 + Arab. 52 + Arab. 53a + Arab. 53b + Arab. 53c + Arab. 53d + Arab. 53e	Recto u. Verso
Aram.	1	NULL	NULL	Vorderseite
G	1289	NULL	NULL	Recto
G	1368	NULL	NULL	Recto
G	1517	NULL	NULL	Vorderseite
G	4737	NULL	NULL	Verso
G	4745	NULL	NULL	Verso
Hebr.	15	NULL	NULL	Vorderseite
Hebr.	17	NULL	NULL	Vorderseite
Hebr.	18	NULL	NULL	Vorderseite
Hebr.	19	NULL	NULL	Vorderseite
Hebr.	20	NULL	NULL	Vorderseite
Hebr.	22	NULL	NULL	Vorderseite
Hebr.	24	NULL	NULL	Vorderseite
Hebr.	25	NULL	NULL	Vorderseite
Kopt.	1207	NULL	NULL	Recto
Kopt.	12	NULL	Frg. 1-7 (Taf. 1-3)	Rückseite
Kopt.	12	NULL	Frg. 1-7 (Taf. 1-3)	Vorderseite
Kopt.	1349	NULL	NULL	Recto
Kopt.	169	NULL	NULL	Recto
Kopt.	452	NULL	NULL	Haarseite u. Fleischseite
Kopt.	685	NULL	NULL	Recto u. Verso
Kopt.	998	NULL	NULL	Recto
```

**Comment by jcowey** [comment]
Thanks, but I am puzzled that "Arab. 1598" = 
http://heidi.ub.uni-heidelberg.de/papyri/default/table/view/papyri/2319
and
http://heidi.ub.uni-heidelberg.de/papyri/default/table/view/papyri/5023
still appears on this list, because 
2319 has "recto" and 5023 has "verso"
Did you use the updated database to do the search? @kadowa :question:

**Comment by kadowa** [comment]
Good point. :flushed: Sorry, I'm switching between different projects and of course forgot to update my local copy of the database. Here is the update list:

```
Arab.	50	NULL	+ Arab. 51 + Arab. 52 + Arab. 53a + Arab. 53b + Arab. 53c + Arab. 53d + Arab. 53e	Recto u. Verso
Aram.	1	NULL	NULL	Vorderseite
G	1289	NULL	NULL	Recto
G	1368	NULL	NULL	Recto
G	4737	NULL	NULL	Verso
G	4745	NULL	NULL	Verso
Hebr.	15	NULL	NULL	Vorderseite
Hebr.	17	NULL	NULL	Vorderseite
Hebr.	18	NULL	NULL	Vorderseite
Hebr.	19	NULL	NULL	Vorderseite
Hebr.	20	NULL	NULL	Vorderseite
Hebr.	22	NULL	NULL	Vorderseite
Hebr.	24	NULL	NULL	Vorderseite
Hebr.	25	NULL	NULL	Vorderseite
Kopt.	1207	NULL	NULL	Recto
Kopt.	12	NULL	Frg. 1-7 (Taf. 1-3)	Rückseite
Kopt.	12	NULL	Frg. 1-7 (Taf. 1-3)	Vorderseite
Kopt.	1349	NULL	NULL	Recto
Kopt.	169	NULL	NULL	Recto
Kopt.	452	NULL	NULL	Haarseite u. Fleischseite
Kopt.	685	NULL	NULL	Recto u. Verso
Kopt.	998	NULL	NULL	Recto
```

**Comment by jcowey** [comment]
Many thanks that list looks much better!! @kadowa

**Comment by jcowey** [comment]
Kopt. 1207 (id: 9901) has been changed to Kopt. 1200 for which there was up til now NO entry. This was a mistake carried over from the original FM database, where there was no entry for Kopt. 1200 (there were two records for 1207, mistakenly)

**Comment by jcowey** [comment]
Kopt. 12. We created two new records (id: 20774; 20775) to accommodate Tafel 3 which we called P.Kopt. 12 c. a and b we introduced to distinguish between Tafel 1 (id:8211; 10087) and Tafel 2 (id:10091; 10092). Care will be needed when matching these to the digital images. Papyrologists should check the two records for "c" to make sure that the entries are correct.

**Comment by jcowey** [comment]
Kopt. 452 (id: 8878) => Haarseite. id:10095 => Fleischseite. Done arbitrarily. Two pages in a codex. Both pages have numbering. Impossible to tell at a glance whether both sides on the front are really both Haarseite.

**Comment by jcowey** [comment]
Hebr. 24 (id: 19485 and 19502) have been deleted. They were duplicates of (id: 5008 and 5009, both from P.Heid. Arab. database), which entered the database from P.Heid. Hebr. database.

**Comment by jcowey** [comment]
Hebr. 22 (id: 19483) has been deleted.

**Comment by jcowey** [comment]
Hebr. 20 (id: 19481) has been deleted. 
At this stage there are a total of 15285 records in the new Database.

**Comment by jcowey** [comment]
Hebr. 19 (id: 19480) has been deleted.
At this stage there are a total of 15284 records in the new Database.

**Comment by jcowey** [comment]
Hebr. 18 (id: 19479) has been deleted.
At this stage there are a total of 15283 records in the new Database.

**Comment by jcowey** [comment]
Hebr. 1 (id: 19478 and 19508) have been deleted.
At this stage there are a total of 15281 records in the new Database.
Der Aufkleber "Rückseite ist P Heid Inv Arab 917" wird entfernt.

**Comment by jcowey** [comment]
Hebr. 25 (id: 19486 and 19501) have been deleted.
At this stage there are a total of 15279 records in the new Database.

**Comment by jcowey** [comment]
Hebr. 15 (id: 19476) has been deleted.
At this stage there are a total of 15275 records in the new Database.

**Comment by jcowey** [comment]
@kadowa 
You might like to update the database and create a final list of the non-unique identifiers based on
`p_heid_inv p_heid_inv_nr p_heid_inv_buchstabe rv p_heid_inv_erg`
There should I hope be just nine left. We aim to have them done by Wednesday of this week. The remaining ones are pretty horrible and tricky.

**Comment by kadowa** [comment]
It's exactly the ones left unchecked on the list above:

```
Arab.	50	NULL	+ Arab. 51 + Arab. 52 + Arab. 53a + Arab. 53b + Arab. 53c + Arab. 53d + Arab. 53e	Recto u. Verso
Aram.	1	NULL	NULL	Vorderseite
G	1289	NULL	NULL	Recto
G	1368	NULL	NULL	Recto
G	4737	NULL	NULL	Verso
G	4745	NULL	NULL	Verso
Kopt.	169	NULL	NULL	Recto
Kopt.	685	NULL	NULL	Recto u. Verso
```

**Comment by jcowey** [comment]
id:10122 was deleted (duplicate of id:8433). Total entries now 15274.

**Comment by jcowey** [comment]
id: 5020 was deleted. Total entries now 15273. Each of 50, 51, 52, 53a, 53b, 53c, 53d, 53e has the entry Recto u. Verso. That is correct.

**Comment by jcowey** [comment]
id:8192 (duplicate of 4977) was deleted. Total entries now 15272.

**Comment by jcowey** [comment]
id:10944 was deleted (duplicate of 12873). Total entries now 15271.

**Comment by jcowey** [comment]
17336 was deleted (duplicate of 12972). Total entries now 15270.

**Comment by jcowey** [comment]
id:10945 was deleted (duplicate of 16980). Total entries now 15269.

**Comment by jcowey** [comment]
id:10946 was deleted (duplicate of 16993). Total entries now 15268.

**Comment by jcowey** [comment]
@kadowa 
Should now be done!! :) :)

**Comment by kadowa** [comment]
:+1:

**Comment by kadowa** [comment]
Working on #20 I found a number of dubious entries, where the data in `p_heid_inv_buchstabe` looks like it should actually be in `p_heid_inv_erg`:

```
# id, p_heid_inv, p_heid_inv_nr, p_heid_inv_buchstabe, p_heid_inv_erg
'17012', 'G', '4764', ' + 4765 + 4766 + 4777 + 4779 + 4782c-d + 4783', NULL
'17013', 'G', '4764', ' + 4765 + 4766 + 4777 + 4779 + 4782c-d + 4783', NULL
'17014', 'G', '4765', '+ 4764 + 4766 + 4777 + 4779 + 4782c-d + 4783', NULL
'17015', 'G', '4765', '+ 4764 + 4766 + 4777 + 4779 + 4782c-d + 4783', NULL
'17016', 'G', '4766', '+ 4764 + 4765 + 4777 + 4779 + 4782c-d + 4783', NULL
'17017', 'G', '4766', '+ 4764 + 4765 + 4777 + 4779 + 4782c-d + 4783', NULL
'17031', 'G', '4777', '+ 4764 + 4765 + 4766 + 4779 + 4782c-d + 4783', NULL
'17032', 'G', '4777', '+ 4764 + 4765 + 4766 + 4779 + 4782c-d + 4783', NULL
'17034', 'G', '4779', '+ 4764 + 4765 + 4766 + 4777 + 4782c-d + 4783 ', NULL
'17037', 'G', '4783', '+ 4764 + 4765 + 4766 + 4777 + 4779 + 4782c-d ', NULL
'17038', 'G', '4783', '+ 4764 + 4765 + 4766 + 4777 + 4779 + 4782c-d ', NULL
```

**Comment by jcowey** [comment]
@kadowa made a separate issue to deal with this #24.

**Comment by kadowa** [comment]
Final question on this issue (hopefully). After studying `p_heid_inv_erg` in more detail I am convinced it should not become part of the inventory number – in almost all cases, it seems to be information on related papyri or fragments. I made another list of unique identifiers *without* `p_heid_inv_erg` and it looks pretty good. There are only a few that are not unique and I'm almost certain that this is due to some data inconsistencies. Here is a list of the duplicates (`page` is the new field encoding recto (f=front) and verso (b=back).

```
Arab.	7	NULL	f
Aram.	1	NULL	b
G.	14	NULL	f
G.	1744	NULL	f
G.	1752	NULL	f
G.	216	NULL	f
G.	22	NULL	f
G.	295	NULL	b
G.	310	NULL	f
G.	32	NULL	f
G.	81	NULL	f
G.	87	NULL	fb
G.	8	NULL	f
Hebr.	1	NULL	b
Hebr.	1	NULL	f
Hebr.	23	NULL	b
Hebr.	23	NULL	f
Hebr.	3	NULL	f
```

I took a quick look at some of these, e.g. *Aram 1* – and to me as a non-expert it looks as if someone made a duplicate entry without checking first, if a previous entry existed. The entries differ very slightly sometimes, but not significantly (again, non-expert look).
